Add tables of noise types to noise documentation page (#129)
ihmeuw/pseudopeople: docs/source/noise/index.rst

introduced into the data and may also be called "corruption" or "distortion." By
default, Pseudopeople applies noise to the simulated datasets using some
reasonable settings. If desired, the user can change the noise settings through

> the configuration system---see the `Configuration section <https://pseudopeople.readthedocs.io/en/latest/configuration/index.html>`_ 

Reviewer: ```suggestion
the configuration system---see the :ref:`Configuration section <configuration_main>` 
```

Using `ref` for links within the docs means that it will be a local link if building locally, the URL will automatically change if we move the page, etc.

#. **Row-based noise:** Errors in the inclusion or exclusion of entire rows of
   data, such as duplication or omission

> #. **Column-based noise:** Errors in data entry for an individual field within a

Reviewer: I think I'm going to revise this further in my PR, but it's fine for now.

Reviewer 2: Yeah, I think we need to edit these definitions, but we can do it in a later PR. It occurred to me that in addition to inclusion or exclusion of rows, we could in the future have row-based noise that works similarly to column-based noise, where we select some rows at random or based on certain criteria, and then noise some of the fields in those specific rows. So probably the definitions for row-based and column-based noise should be more symmetrical. But again, that's for a later PR, just writing my thoughts here.

Available Noise Types
---------------------

The table lists all the available noise types, but not every
type of noise will necessarily be applied to every dataset or every column. The
configuration determines which noise types are actually used. The "Noise Type"
column shows the name of the noise type in the configuration system.

.. list-table:: Types of row-based noise (``row_noise``)
  :widths: 1 2 5 
  :header-rows: 1

  * - Noise Type
    - Config key 
    - Example cause
  * - Omit a row
    - ``omit_row``
    - Neglecting to file a tax form on time

.. list-table:: Types of column-based noise (``column_noise``)
  :widths: 1 2 5 
  :header-rows: 1

  * - Noise Type
    - Config key
    - Example cause
  * - Leave a field blank
    - ``leave_blank``
    - Forgetting to write your name on the designated line
  * - Choose the wrong option
    - ``choose_wrong_option``
    - Marking the "Male" box when you meant "Female"
  * - Optical character recognition (OCR) error
    - ``make_ocr_errors``
    - Misreading an 'S' instead of a '5'

>   * - Make typos

Reviewer 2: If we're sticking with listing things in the order of application, typos should be last. Here's the order we specified in the [concept model](https://vivarium-research.readthedocs.io/en/latest/models/concept_models/vivarium_census_synthdata/concept_model.html#noise-functions), with the ones we've implemented in **bold** (and using the old names, not the new names): 

1. **“Borrowed” SSN (happens in simulation NOT noise functions)**
3. **Omissions**
5. Duplications
7. **Missing Data**
9. **Incorrect Selection**
11. Copy from Within Household
13. Month and day swaps (applies to dates only)
15. **Zip Code Miswriting (applies to Zip Code only)**
17. **Age Miswriting (applies to age only)**
19. **Numeric miswriting**
21. Nicknames
23. **Fake names**
25. Phonetic
27. OCR
29. **Typographic**

We talked about re-arranging things a bit, e.g.,

1. **“Borrowed” SSN (happens in simulation NOT noise functions)**
3. **Omissions**
5. Duplications
7. **Missing Data**
9. **Incorrect Selection**
11. Copy from Within Household
21. Nicknames
23. **Fake names**
13. Month and day swaps (applies to dates only)
19. **Numeric miswriting**
15. **Zip Code Miswriting (applies to Zip Code only)**
17. **Age Miswriting (applies to age only)**
25. Phonetic
27. OCR
29. **Typographic**

Or something like that. Should we choose an order and stick with it? Should we document the changes in the concept model (perhaps with our updated naming conventions as well), or is there a better place?

Reviewer: I think we should use the re-arranged order you just suggested in this table, and then once we've filled in the details for all noise types, re-arrange the sections into that order.

Also, just to note -- "happens in simulation NOT noise functions" is not going to be meaningful to our users. Instead we should say it is "not configurable."

Reviewer: Wait actually -- shouldn't "age miswriting" aka "age misreporting" happen before numeric miswriting? I think they both apply to the age column, so it is a meaningful difference.

Reviewer 2: According to the concept model, we didn't ask for numeric miswriting noise on age:
<img width="593" alt="image" src="https://user-images.githubusercontent.com/16144326/234171406-8df8a792-ae2a-441c-b991-77db3c20f71c.png">
Similarly, zipcodes get zipcode miswriting but not basic numeric miswriting. I think the idea was that these specialized functions would replace numeric miswriting for these particular columns. But I don't necessarily see a reason we should _prevent_ a user from applying both types of noise (age + numeric or zip + numeric) if they want to. In which case I think you're right that it probably makes more sense to do the specialized function first and numeric miswriting after.

    - ``make_typos``
    - Accidentally typing an 'l' instead of a 'k' because they are 
      right next to each other on a QWERTY keyboard
  * - Misreport age
    - ``misreport_age``
    - Reporting that you are 28 years old when you are actually 27
  * - Write the wrong digits
    - ``write_wrong_digits``

>     - Writing "2022" instead of "2002" in your date of birth

Reviewer 2: In terms of real-world examples, it seems like this might fall more in the category of "typo" than randomly getting the digits wrong. Would a better example be getting a street number or unit number wrong, e.g., "732" instead of "932", or something that's more like just remembering the number wrong in a somewhat random way?

Reviewer: Yeah, I see what you mean, though it can get a little tricky -- for example if you wrote 1997 instead of 1999 that couldn't be a typo (in our current approach) since 7 and 9 are not next to each other.

I like the idea of changing this example to be about street number though; I think that is a more believable example.

```suggestion
    - Writing "732 Main St" as your street address instead of "932 Main St"
```

  * - Write the wrong zipcode digits 
    - ``write_wrong_zipcode_digits``

>     - Writing ZIP code 98118 when you actually live in 98112

Reviewer: Yeah it might be a bit confusing as-is but I don't see a good way to explain this without adding another column to this table. Especially when we link from this table directly to the detail sections, a reader would only be one click away from the explanation.

>     - Writing ZIP code 98118 when you actually live in 98112

Reviewer 2: > Especially when we link from this table directly to the detail sections

Oh, is that the plan? E.g., turn the names of the noise functions into hyperlinks to the appropriate sections? That sounds like a good idea. In that case, we should probably at least add a note to the description of "write wrong zipcode digits" highlighting how it's different from "write wrong digits," and maybe we should add a note to "write wrong digits" as well.

>   * - Fake names

Reviewer 2: I think the name of this noise type should be "Use a fake name." Does that sound good to people?

Available Noise Types
---------------------

The table lists all the available noise types, but not every
type of noise will necessarily be applied to every dataset or every column. The
configuration determines which noise types are actually used. The "Noise Type"
column shows the name of the noise type in the configuration system.

.. list-table:: Types of row-based noise (``row_noise``)
  :widths: 1 2 5 
  :header-rows: 1

  * - Noise Type
    - Config key 
    - Example cause
  * - Omit a row
    - ``omit_row``
    - Neglecting to file a tax form on time
  * - Duplicate a row
    - ``duplicate_row``
    - College student being reported both at their university and their guardian's
      home address

.. list-table:: Types of column-based noise (``column_noise``)
  :widths: 1 2 5 
  :header-rows: 1

  * - Noise Type
    - Config key
    - Example cause
  * - Leave a field blank
    - ``leave_blank``
    - Forgetting to write your name on the designated line
  * - Choose the wrong option from a fixed set of options
    - ``choose_wrong_option``
    - Marking the "Male" box when you meant "Female"

>   * - Optical character recognition (OCR) error

Reviewer: I'm wondering what we should do about these, which we won't have for CODS? @aflaxman any thoughts, should we add an asterisk or something? Remove them altogether for now?

Author: Oh I didn't realize these hadn't been implemented yet. If we want to follow same thing we did for the datasets page, I could add a 'coming soon' statement to the noises that haven't been added yet (i.e., like what I said for the 1040 tax forms).

Reviewer 2: As discussed in the documentation session, I think we should remove the unimplemented things for now (but save your work somewhere if you've already written something!)

I don't think OCR noise is implemented yet.

    - ``make_ocr_errors``
    - Misreading an 'S' instead of a '5'
  * - Phonetic error
    - ``make_phonetic_errors``
    - Mishearing a 't' for a 'd'
  * - Typographic errors
    - ``make_typos``
    - Accidentally typing an 'l' instead of a 'k' because they are 
      right next to each other on a QWERTY keyboard
  * - Nicknames
    - ``use_nickname``
    - Writing 'Alex' instead of legal name 'Alexander'
  * - Fake names
    - ``use_fake_name``

>     - Using fake name rather than official name

Reviewer 3: ```suggestion
    - Using 'Mr' rather than actual first name
```

  * - Numeric miswriting 
    - ``write_wrong_digits``

>     - Listing incorrect date of birth 

Reviewer: ```suggestion
    - Writing "2022" instead of "2002" in your date of birth
```

>   * - Age miswriting

Reviewer: I don't think this name matches with what is on the noise types page, can you check these?

    - ``misreport_age``

>     - Listing incorrect age

Reviewer: ```suggestion
    - Reporting that you are 28 years old when you are actually 27
```

  * - ZIP code miswriting
    - ``write_wrong_zipcode_digits``

>     - Listing incorrect ZIP code

Reviewer: ```suggestion
    - Writing ZIP code 98118 when you actually live in 98112
```

  * - Copy from within household
    - ``copy_from_household_member``

>     - Copying survey responses from another household member

Reviewer: ```suggestion
    - Accidentally writing the age of another person in your household in the line for your age
```

  * - Month and day swap
    - ``swap_morth_and_day``

>     - Reporting date as DD/MM/YYYY instead of MM/DD/YYYY

Reviewer: ```suggestion
    - Reporting 17/05/1976 when a survey asks for the date in MM/DD/YYYY format
```
